File: zig/rna_transcription.zig
const mem = @import("std").mem;

pub fn toRna(allocator: mem.Allocator, dna: []const u8) mem.Allocator.Error![]const u8 {
    const rna = try allocator.alloc(u8, dna.len);
    for (dna, 0..) |d, i| {
        rna[i] = switch (d) {
            'G' => 'C',
            'C' => 'G',
            'T' => 'A',
            'A' => 'U',
            else => unreachable,
        };
    }
    return rna;
}

File: ocaml/rna_transcription.ml
type dna = [ `A | `C | `G | `T ]
type rna = [ `A | `C | `G | `U ]

let rec to_rna = function
  | [] -> []
  | `G :: ds -> `C :: to_rna ds
  | `C :: ds -> `G :: to_rna ds
  | `T :: ds -> `A :: to_rna ds
  | `A :: ds -> `U :: to_rna ds
  | _ -> failwith "error"
